[New Feature] I2C position encoder support

bgort opened the pull request:
This is an initial commit related to I2C position encoder support.  

This is being PR'd primarily to replace #6675, which targeted the wrong branch.

It does compile (with a few warnings) though it has not yet been tested.

I plan to continue improving/cleaning this up, as there are areas that need considerable work.

This is -not- ready to be merged.

bgort committed:
[New Feature] I2C position encoder support

This is an initial commit, and while it does compile (with a few warnings)
it has not yet been tested.

This is being PR'd primarily to replace #6675, which targeted the wrong branch.

I plan to continue improving/cleaning this up, as there areas that need work.

This is -not- ready to be merged.

bgort committed:
let the cleanups begin.

bgort committed:
progress

bgort committed:
more progress

bgort committed:
comments, rename files, etc.

bgort committed:
clean

bgort commented:
This has been significantly cleaned up, optimized, etc. and now compiles cleanly.  

It still needs more time spent on it, but is probably okay enough for inclusion after some basic testing to make sure there are no obvious issues.

My encoder modules arrived yesterday, so I'll do my best to test this week.

bgort commented:
Currently, this adds about 12k to PROGMEM and uses ~500 bytes of SRAM?  A good portion of the SRAM usage is error correction-related, it appears, and I think that can be significantly improved.

thinkyhead commented on Marlin/gcode.h:
```diff
@@ -139,6 +149,17 @@ class GCodeParser {
     }
 
+    static volatile bool seen_any() {
+      return strlen(command_args)!=0; // todo: is this right?
```
This will be more concise:
```cpp
return *command_args == '\0';
```

bgort replied:
Yeah, that makes good sense -- I just didn't know how command_args was left, so wasn't sure of the best way to detect 'empty', hence the 'is this right?' ...

Thanks!

thinkyhead commented on Marlin/gcode.h:
```diff
@@ -128,6 +128,16 @@ class GCodeParser {
+      return TEST(codebits[23 >> 3], 23 & 0x7) || //X
+             TEST(codebits[24 >> 3], 24 & 0x7) || //Y
+             TEST(codebits[25 >> 3], 25 & 0x7) || //Z
+             TEST(codebits[4 >> 3], 4 & 0x7); //E
```
For concision and to make the meaning more clear…
```cpp
#define SEEN_TEST(L) TEST(codebits[(L - 'A') >> 3], (L - 'A') & 0x7)
return SEEN_TEST('X') || SEEN_TEST('Y') || SEEN_TEST('Z') || SEEN_TEST('E');
```

bgort replied:
Makes good sense.  👍

thinkyhead commented on Marlin/gcode.h:
```diff
@@ -139,6 +149,17 @@ class GCodeParser {
+             !!strchr(command_args, 'Y') || //Y
+             !!strchr(command_args, 'Z') || //Z
+             !!strchr(command_args, 'E'); //E
+    }
```
For concision and to make the meaning more clear…
```cpp
#define SEEN_TEST(L) !!strchr(command_args, L)
return SEEN_TEST('X') || SEEN_TEST('Y') || SEEN_TEST('Z') || SEEN_TEST('E');
```
Since the `seen_any_axis` function will be the same in both cases, it can be defined just once, below.

thinkyhead commented on Marlin/gcode.h:
```diff
@@ -148,6 +169,9 @@ class GCodeParser {
   FORCE_INLINE static bool has_value() { return value_ptr != NULL; }
 
+  // Seen and has value
+  FORCE_INLINE static bool seenval(const char c) { return seen(c) && has_value(); }
```
`has_value()` can't return `true` unless `seen` is also true, so this is just redundant. The `has_value` method already covers this case.

bgort replied:
`has_value()` doesn't take an argument, though.

bgort replied:
Also, I anticipate replacing at least a ~12-18~ quite-a-few-more `seen(X) && has_value()` conditionals scattered throughout the codebase.  Didn't include them here because they're unrelated.

bgort replied:
For instance...

```
g26_repeats = parser.seen('R') ? (parser.has_value() ? parser.value_int() : GRID_MAX_POINTS + 1) : GRID_MAX_POINTS + 1;
```
becomes
```
g26_repeats = parser.seenval('R') ? parser.value_int() : GRID_MAX_POINTS + 1);
```

thinkyhead replied:
Just be careful. The `g26_repeats` code is already badly written, so the fix there is obvious. Generally speaking we should be flagging an error when parameters that require a value are supplied without a value. Ignoring the parameter when it lacks a value might not be the expected behavior in all cases. So… just be selective in its use.

bgort replied:
Yes, of course.  There are places it makes sense, and places where things should be handled better/differently...

thinkyhead commented on Marlin/macros.h:
```diff
@@ -108,6 +108,8 @@
 #define HYPOT2(x,y) (sq(x)+sq(y))
 #define HYPOT(x,y) sqrt(HYPOT2(x,y))
 
+#define SIGN(a) ((a>0) - (a<0))
```
This should be either: `((a > 0) - (a < 0))` or `((a>0)-(a<0))`, whichever is most preferred.

thinkyhead commented on Marlin/enum.h:
```diff
@@ -34,23 +34,29 @@
+#define LOOP_LE_N(VAR, N) LOOP_S_LE_N(VAR, 0, N)
+#define LOOP_L_N(VAR, N) LOOP_S_L_N(VAR, 0, N)
+
+#define LOOP_NA(VAR) LOOP_L_N(VAR, NUM_AXIS)
```
At this point we don't need this, but it's forward-thinking. There may be occasions where another coordinated axis is wanted, probably inserted between Z and E.

bgort replied:
Yeah, they're currently unused, but I could see a future use so left them in.  (I used most of them briefly, then went another route.)

bgort committed:
Cleanups per thinkyhead

bgort committed:
a few more cleanups

thinkyhead commented:
If it's using up a large amount of SRAM, first thing to check whether it has C style strings that are not assigned with PSTR or PROGMEM.

bgort commented:
Yeah, there was a lot of that kind of thing, but @jaysonkelly (the original contributor) fixed most of it and I've been through it all a few times and don't see any more.  I plan to go through it a few more times to see what I can see, and if I come up with nothing I'll probably do a deeper analysis with objdump, etc.

Last I checked it was using 458, while the variables in the classes suggest that about half that should be used (unless I added wrong).  Will sort it out.

jaysonkelly commented:
@bgort thanks for all the development so far! Let me know when you get your printer set up. I suggest starting with the x and y axes. Let me know if you have any questions.

bgort commented:
No problem.

And yeah - thank you.  I'll probably start with my X axis alone, as that'll be easiest to work with on my setup.  I'll be doing reliability and accuracy testing straight away (with a micron dial indicator), then I'll probably go on from there to implement 'mechanical endstop replacement mode' by detecting the end of the strip, etc.  Then I want to sort out leveling of the two sides of my Z axis by independently moving one of the Z steppers.

And I also want to replace counting steps, etc. with reading the encoders to determine position if they do what they're supposed to do (which I assume they will).

bgort commented:
In case anyone who doesn't already know this cares:

This will give you a sorted list of variables/objects using SRAM:
` avr-nm -Crtd --size-sort <your elf file> | grep -i ' [dbv] '`

Pretty interesting/useful, generally.

EDIT:

Here's some output:
```
[root@create][628][BGMarlin]# avr-nm -Crtd --size-sort .pioenvs/rambo/firmware.elf | grep -i ' [dbv] '
00001296 B Planner::block_buffer
00000900 B unified_bed_leveling::z_values
00000384 B command_queue
00000256 b u8g_dev_st7565_nhd_c12864_2x_buf
00000245 B I2CPEM
00000130 B rx_buffer
00000108 B _ZL10servo_info.lto_priv.164
00000096 D _ZZL16pin_is_protectedhE14sensitive_pins.lto_priv.207
00000096 b get_serial_commands()::serial_line_buffer
00000067 D lcd_status_message
00000062 B u8g
00000036 b screen_history
00000032 B TwoWire::txBuffer
00000032 B TwoWire::rxBuffer
00000032 b horizontal_mesh_line_flags
00000032 b vertical_mesh_line_flags
00000032 B _ZL12circle_flags.lto_priv.214
00000032 b twi_txBuffer
00000032 b twi_rxBuffer
00000032 B twi_masterBuffer.lto_priv.158
00000030 B buzzer
00000026 B GCodeParser::param
00000024 B hotend_offset
...
```

Memory usage is in the leftmost column, in bytes.

bgort committed:
cleanups, bugfixes, etc.

bgort committed:
remove unnecessary passes_test(), additional cleanups/optimizations

jaysonkelly commented:
Now that you have been working with the code, do you see anything that might cause the secondary axis (Y in my case) to stop correcting? It seems almost random, but I've had two.printers fail to correct in the middle of the print while the encoders should have been active.

bgort commented:
No, not yet, but perhaps it'll sort itself out.  I have seen a few bugs in a few places and corrected them, and I've been optimizing a lot, so maybe I got it.  If not, we'll definitely find whatever it is eventually.  I plan to properly abuse everything.

Right now, I'm having an issue with the modules not communicating with Marlin.  They seem to start up fine (green, green, blue) and they detect the strip fine (LED changes color depending on distance, etc.), but there's no response to anything.  Wondering if it's possible the wires were flipped on the connectors (the red and black are backwards everywhere -- red is ground, black is +5V -- which works fine, of course, but had to make sure everything was wired correctly when I noticed the backwardness) and will be getting out the oscilloscope/logic analyzer shortly to see if SCL and SDA are backwards.  If so, that's an easy fix, but haven't had time yet.  Hopefully today.

bgort committed:
cleanups

bgort commented:
Also, not sure if you're aware, but it looks like @chrissbarr has pushed a new encoder board revision to his repo.  It looks a little more robust - has its own regulator, and several connectors for (I presume) daisy-chaining.

For whatever it's worth, I do like your connector better than the smaller ones.  I've tried to use those on a few projects and they're always difficult to make, work with, etc.; a real pain.

jaysonkelly commented:
@bgort Have you set the individual addresses yet? They all start at a default address of 30. The addresses can be changed via the cut traces or by the custom M codes.

bgort commented:
Yeah, I understand.  No, I haven't set addresses yet because Marlin doesn't see the modules at all at the default address.  They're completely unresponsive (as though they aren't hooked up).

jaysonkelly commented:
Maybe there are board settings set on rigidboardv2 and not on yours?

bgort commented:
I'm using a RAMBo, but yeah it's possible something needs to be enabled somewhere, though I don't imagine (hope) it's anything to do with the hardware.  I've never connected anything to the I2C pins on this board, so they should be okay.

I should be able to chase this down today.

jaysonkelly commented:
All the splitter to encoder cables should be wired correctly since they are
straight across and I tested each for continuity. I went with the schematic
online for the board to splitter cable. Is it possible that you have a
different version than what is online? Or maybe I got the sda/sdl swapped.

On Thu, Jun 8, 2017, 6:47 AM bgort <notifications@github.com> wrote:

> I'm using a RAMBo, but yeah it's possible something needs to be enabled
> somewhere, though I don't imagine (hope) it's anything to do with the
> hardware. I've never connected anything to the I2C pins on this board, so
> they should be okay.
>
> I should be able to chase this down today.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/6946#issuecomment-307092598>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATqPZOGV0ZoKbO6cvRRq0ciAQjB9XJHrks5sB-1cgaJpZM4NvJ-_>
> .
>

bgort commented:
There's good continuity between all the wires on all of the connectors -- just verified them all.  

The image online looks consistent with my board.  FWIW, I have a 1.3L.

I think the only remaining possibilities are that SDA/SCL are flipped, something needs to be enabled somewhere, or I have bad I2C pins.  Will check with the logic analyzer shortly.

bgort commented:
Wires were swapped.  Good communication now, so all is well.

Also, I got a new scope/logic analyzer out of all of this.  I was using a red pitaya without a cover and I shorted a few pins while trying to calibrate it (the stored DC offset was wrong), which unfortunately let the smoke out....

bgort committed:
misc.

jaysonkelly commented:
Glad to hear things are communicating! I just compiled the latest build configured for rigidbot_v2 board with good results. The init function had a few old function names in the y z and e sections, but other than that it works.

jaysonkelly commented:
I was able to upload it and find some errors in my configuration before I had to leave it. 

Suggestion: when the printer detects bad magnetic strength for a short time, assume the worst, pause and re-home that axis. What do you think?

bgort commented:
>  Glad to hear things are communicating! I just compiled the latest build configured for rigidbot_v2 board with good results. The init function had a few old function names in the y z and e sections, but other than that it works.

That's good to hear!  I still haven't actually tested anything but I'm getting there.

Not surprised a few of the old function names are stuck in the init -- sometimes find and replace doesn't get everything (when I click the wrong button!) ..

> Suggestion: when the printer detects bad magnetic strength for a short time, assume the worst, pause and re-home that axis. What do you think?

Yeah, I was considering rewriting some of that so it's a little more robust, generally.  Not sure exactly how yet, but rehoming, etc. may end up being the best way.

I tend to test/work starting with the 'easy/simple stuff' and then move into the more complex (and important) .. just haven't gotten past the simple stuff yet.  Been one of those days.

bgort committed:
Fix up I2CPEM.init() and a few other things.

bgort committed:
organize, fix, rename, etc.

bgort committed:
more optimization

bgort commented:
@jaysonkelly Any chance you can give this a quick test this morning?  I still haven't had time to get everything fully tested here, but would like to merge this so I can use a few of the other changes in other parts of the codebase.  

I'm confident this won't break anything as it's all conditionally compiled, so I don't see a problem with merging now.  I do plan to continue working on it regularly.

EDITED:  I accidentally clicked 'Comment' before I was done....

bgort commented:
Think there may be a bug in the encoder module firmware.  Haven't yet had time to go looking for it, but when changing the address via M864 the modules seem to lock up before they do their restart.  The address is successfully changed, but I've had to disconnect and reconnect each module to get it to respond.

@chrissbarr Any chance you'd be willing to put the encoder module firmware in its own repository?  Happy to help out with it, but would be nice for it to be a standalone project rather than a few directories deep.

bgort committed:
a few more tweaks

bgort commented:
Going to go ahead and merge this now.  It seems to be working, as far as I can test it, and it's fully conditionally compiled, so can't break anything else.

jaysonkelly commented:
Fantastic! I've got work and then we are getting ready for a yard sale tomorrow morning, but I'll try and work it in there somewhere before my business trip next week. I'm not really a coder just a mechanical engineer with a knack for figuring things out, so I really appreciate the effort you have put into the code to get it into marlin. Thanks!

bgort commented:
No problem.  I'm having fun working on it...

I plan to do some more cleanups/optimization, and then I think next up is homing using the encoders, though I'm not positive that's possible because I don't yet know how they'll behave when they reach the end of the strip.  Obviously they'll report they've lost signal, but will they lose signal at the same position +/- a micron or two every time?  We shall see.

jaysonkelly commented:
I'm pretty sure I saw something in the encoder data sheet about it
recognizing the end of the strip as distinct to magnetic strength loss.
There may need to be a modification to the firmware on the encoders to take
advantage of it though.

On Fri, Jun 9, 2017, 6:34 AM bgort <notifications@github.com> wrote:

> No problem. I'm having fun working on it...
>
> I plan to do some more cleanups/optimization, and then I think next up is
> homing using the encoders, though I'm not positive that's possible because
> I don't yet know how they'll behave when they reach the end of the strip.
> Obviously they'll report they've lost signal, but will they lose signal at
> the same position +/- a micron or two every time? We shall see.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/6946#issuecomment-307376530>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATqPZIGXAw69u1Xd2L_1st20LPzd62uyks5sCTu6gaJpZM4NvJ-_>
> .
>

bgort commented:
Thanks.  I'm checking this out now.

Also, just to confirm, the ticks per unit with your strip should be 1024?  Reading the datasheet, they're warning against using anything but 1mm pole pair strip, as doing so introduces non-linearities.

Are you willing to share your sources for the strip (so I don't have to go searching), as I may want to try 1mm too, just to satisfy myself.  Appreciate it!

jaysonkelly commented:
That is correct, 1024. Of course! I'm pretty sure they only sell it in bulk
30 foot minimum, but you can get a sample strip for free from atmel. I'll
get you our source when I can find it.

On Fri, Jun 9, 2017, 6:58 AM bgort <notifications@github.com> wrote:

> Thanks. I'm checking this out now.
>
> Also, just to confirm, the ticks per unit with your strip should be 1024?
> Reading the datasheet, they're warning against using anything but 1mm pole
> pair strip, as doing so introduces non-linearities.
>
> Are you willing to share your sources for the strip (so I don't have to go
> searching), as I may want to try 1mm too, just to satisfy myself.
> Appreciate it!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MarlinFirmware/Marlin/pull/6946#issuecomment-307381550>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATqPZKpI725meE9UOHOyCaYAgXUWytHtks5sCUGDgaJpZM4NvJ-_>
> .
>

jaysonkelly commented:
Bogen electronic.

On Fri, Jun 9, 2017, 7:06 AM Jayson Kelly <jaysonkelly@gmail.com> wrote:

> That is correct, 1024. Of course! I'm pretty sure they only sell it in
> bulk 30 foot minimum, but you can get a sample strip for free from atmel.
> I'll get you our source when I can find it.
>
> On Fri, Jun 9, 2017, 6:58 AM bgort <notifications@github.com> wrote:
>
>> Thanks. I'm checking this out now.
>>
>> Also, just to confirm, the ticks per unit with your strip should be 1024?
>> Reading the datasheet, they're warning against using anything but 1mm pole
>> pair strip, as doing so introduces non-linearities.
>>
>> Are you willing to share your sources for the strip (so I don't have to
>> go searching), as I may want to try 1mm too, just to satisfy myself.
>> Appreciate it!
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/MarlinFirmware/Marlin/pull/6946#issuecomment-307381550>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ATqPZKpI725meE9UOHOyCaYAgXUWytHtks5sCUGDgaJpZM4NvJ-_>
>> .
>>
>

bgort commented:
Appreciate it.  Didn't know Atmel sold that sort of thing; will look into.  Thanks!

bgort commented:
> Bogen electronic.

Great!

bgort commented:
Do you know offhand the part number or specs of what you sent me?  I'm mostly interested in the accuracy class (A03-A100), but whatever other details you have would be good to know.

thinkyhead commented:
What happened to keeping this as a PR? It's not really done. It ended up getting direct-committed to the `bugfix-1.1.x` branch, kind of lousing up the history if we like tracking PRs.

bgort commented:
It's done enough to be usable - basic functionality - and there are no errors or other issues.  It wasn't direct committed, but squashed and merged as a PR.

thinkyhead commented:
Well, among other things, when you make a change to the main `Configuration.h` or `Configuration_adv.h` file, you must propagate the same changes to all the example configurations. If this had received a proper review that would have been required before any merge.

bgort commented:
Of course.  I actually meant to do that but it slipped my mind.  Will take care of it now.

thinkyhead commented:
Then there's stuff like…
```cpp
lastUpdateMillis + I2CPE_MIN_UPD_TIME_MS) < millis()
```
Please use the `ELAPSED` / `PENDING` macros. They exist specifically to ensure proper handling of time intervals. And use "next" not "last" variables to track timeouts.

thinkyhead commented:
Otherwise, the PR is fine. These are some of the things I need to include in the documentation.

bgort commented:
There are a number of things that still need to be optimized, obviously, but it does work as it is.  

> Please use the ELAPSED / PENDING macros. They exist specifically to ensure proper handling of time intervals.

I'd have happily changed that, too, if you'd mentioned it when you made comments the other day.  How are we to know when a 'proper review' has taken place?

Not sarcastic - serious question:  Would you prefer I just treat this as your project and not merge things until you've said they're okay?  There was no discussion about anything, so I'm doing things the way I would do them for myself.  If that isn't going to work for you, just let me know.

thinkyhead commented:
I've been too busy to attend to as many issues as I would like, otherwise I would have provided the additional feedback. I may presume that regular contributors already understand that all the example configurations need to be kept up to date with new changes, so I may not always mention it.

New PRs that add features and make changes to the configurations are _always_ going to require at least one major revision. Either I'll make a copy of the PR, clean it up and merge it —giving a chance to clean up language translation issues, spelling, grammar, and spacing, ensure that it's compliant and takes advantage of facilities provided— or after the PR is merged I will follow it up with the same corrections in a new PR.

We don't have a policy that you cannot merge your own PRs, mainly because I am constantly merging my PRs without any fanfare, just to get them tested, and I don't want to hypocritically flout the privilege while insisting that no one else can do so. After all, there are times when it makes sense to push through a bug-fix quickly, even if it isn't the most perfect jewel.

So, I've been thinking about this question to some extent. I think that for patches and incremental improvements, we can have some leeway, but for new features and things involving new systems, long passages of documentation, etc., it should come down to an "Executive Editor" to review the PR, decide when it's ready, and to do the actual merge.

While the de-centralized nature of open source and Github is awesome, and the egalitarian meritocratic nature of the team aspect of open source is also awesome, I think it is important to have a single executive who keeps the whole picture in mind and considers contributions with respect to the future roadmap of the project. It lends a central vision to the project, and if the project executive can make regular blog posts and share the project vision and roadmap with the world, all the better.

I've been taking on some of that role in handling the Twitter for [@MarlinFirmware](https://twitter.com/MarlinFirmware), writing a lot of the website, managing releases, and writing semi-regular updates on my Patreon page, but haven't formally adopted an executive role in the project itself.

A lot of my week is taken up working for a startup in Austin, TX, or I would give more feedback on pending PRs. For the time being, please let them ferment a bit longer. I will try not to let PRs linger for too long, but note that my emphasis is not on merging new features right now, but on cleanup and bug fixes. I very much want to move forward with 1.2, the new file hierarchy, and the HAL. And I might have even wanted to hold off on this new non-essential feature until after then. But I'm waiting for things to slow down in the issue queue and make sure that 1.1.x is solid before leaping forward with 1.2.

There are now 4 of us with merge privileges. Perhaps with our combined energy we can patch the bugs, fix the issues, clean up the code, and make sure everything works nicely while holding back new features and contributions and putting off everyone so that we can have a nice slow point and make the transition to 1.2.

chrissbarr commented:
@bgort I've moved the encoder firmware to a standalone repository [here](https://github.com/Aus3D/Magnetic-Encoder-Firmware). Hopefully that makes the code a tad easier to work with - I can understand not wanting to pull a repo full of CAD files to tweak a few lines of code. If there's anything else I can do to make it easier to work with just let me know.

Happy to see any PRs for bugs you find, from what you describe it sounds like the WDT isn't resetting the modules and I'm not sure what would be causing that. There have been a few changes to the code since I last thoroughly tested things, but nothing (I thought) that touched on that area of things.

Just want to say I really appreciate the work you've put in cleaning this up and getting it integrated into Marlin nicely! Seeing the progress you were making was my motivation to start fleshing out the v2 modules, so many thanks for that. Hopefully, soon there will be some new hardware to test your improvements with!

My supplier for one-off / small quantities of the magnetic strips during testing was through AMS [here](http://ams.com/eng/Products/Magnetic-Position-Sensors/Magnets/AS5000-MS10-300). AFAIK they're just short pre-cut lengths of a strip from Bogen Electronic. They have a datasheet that covers most of the strip's specifications on that page, though I'm not sure if they specifically state the accuracy class (they seem to list a few options).

Strip with 1mm poles / 2mm pole-pair length is all that I've tried, as it seems to be what's recommended in the datasheet. I've gotten pole-length and pole-pair length mixed up a few times, so all I'd say is double check which numbers are being talked about - I didn't realise @jaysonkelly and co were actually using a strip of a different density, but if 1024 ticks/mm is working for them then they must be.

Anyway, let me know if there's anything else I can do to help here. Otherwise, I am watching keenly!

thinkyhead commented:
@bgort One thing I would certainly have mentioned, had this been held longer, is that you cannot use `N` as a G-code parameter. You'll have to change `M864` to use another letter. Please also avoid `M` and `G`.

thinkyhead commented:
@bgort We will also need to make `I2CPositionEncoder` into a static singleton by making all its data and methods `static`.

bgort commented:
> I've been too busy to attend to as many issues as I would like, otherwise I would have provided the additional feedback. I may presume that regular contributors already understand that all the example configurations need to be kept up to date with new changes, so I may not always mention it.

I understand, and yes, I'm aware of the need for and the reasoning behind keeping example configs current (I've made changes to all example configs 2-3 times in the past).  As I said, I had planned to do it this time but it just slipped my mind this morning, like anything slips anyone's mind.  A simple 'looks like you forgot the example configs' would have been sufficient, and I would have laughed, been embarrassed that I forgot them, fixed it straightaway, and moved on.

> New PRs that add features and make changes to the configurations are always going to require at least one major revision.

Of course.  And I expect many more major and minor revisions for this one.  There's still a good bit that needs to be optimized, cleaned up, improved, etc.  And there's a lot of functionality I'd like to add, too.  

I chose to merge this now because it's in a working state, compiles cleanly, because merging allows others to more easily test (as you pointed out) and contribute/add to/expand on, and because I didn't see a reason not to include it.  (To be completely transparent, I also wanted to merge it because I wanted to make use of some of the things I added in other parts of the codebase, -and- because it's annoying to keep rebasing, etc.)  As I mentioned, it's fully conditionally compiled, so unless someone enables it intentionally they won't even notice it's there.  

To be perfectly clear:  I wasn't aiming for perfect, but for 'good enough for initial inclusion', and I do believe it is 'good enough' for an initial capability.  (And I'm not criticizing [I mean, **it happens], but a lot of your merges contain bugs/issues, too, so...)

> Either I'll make a copy of the PR, clean it up and merge it —giving a chance to clean up language translation issues, spelling, grammar, and spacing, ensure that it's compliant and takes advantage of facilities provided— or after the PR is merged I will follow it up with the same corrections in a new PR.

This is what I expected - a followup PR that made the corrections you felt were necessary, or just some simple feedback that I would have happily considered and very likely implemented.  

FYI, you can trust that I'm not ever going to knowingly merge something that is going to cause problems elsewhere or that I haven't tested (at least to the extent I reasonably can).  As far as I know, I haven't yet introduced a single breaking change, and I don't plan to start anytime soon.  I wrote and currently develop/manage a codebase about 3x the size of Marlin, and I'm pretty good about not breaking things before pushing them; if I wasn't, I'd have hundreds of complaints that I'd have to deal with (which is what happened the last time I broke something - ~2 years back; it took me nearly a full day to dig out of the support emails).  Anyway, my point is that I'm generally extremely careful about making changes that can affect users - and my caution, the amount of testing, etc. scales with the scope of the changes and the potential magnitude of the effects.

> So, I've been thinking about this question to some extent. I think that for patches and incremental improvements, we can have some leeway, but for new features and things involving new systems, long passages of documentation, etc., it should come down to an "Executive Editor" to review the PR, decide when it's ready, and to do the actual merge.

I think this makes good sense, generally but I don't think there should be a single 'executive editor'.  

Instead, I think it makes sense that all major changes (major new features, like this I2CPE stuff) should be reviewed by someone (anyone) else with merge privileges, and they can then provide feedback, etc., and do the merge when they're satisfied.  I've suggested something similar on a few occasions.

Having a single executive will likely i) overburden that volunteer person, ii) slow the pace of development, iii) change the tenor/tone/feel of the project, and iv) be problematic in other ways.  It doesn't feel good to me.

> While the de-centralized nature of open source and Github is awesome, and the egalitarian meritocratic nature of the team aspect of open source is also awesome, I think it is important to have a single executive who keeps the whole picture in mind and considers contributions with respect to the future roadmap of the project. It lends a central vision to the project, and if the project executive can make regular blog posts and share the project vision and roadmap with the world, all the better.

I really like the idea of having a roadmap, regular open discussions about direction/architecture/etc., and more structure, generally, but I'm entirely uncomfortable following someone else's lead.  I'm certainly willing to defer to you and Roxy on what works best, fits best within the project, etc. because you have more experience than I do right now, but that's different than an 'executive' setting down priorities and a plan and having to follow it if you want your changes incorporated.

I think the central vision can and should be defined by the core group of developers, and ultimately by the users of the project through feedback, discussion, etc.

Also, I've had the feeling on a few occasions that you may be wanting to take Marlin in more-business-like direction either personally or for the project (maybe like many of the projects that leave their codebase open source but offer services layered on top, or something along those lines), so given that what you've said here triggered those feelings again, it's probably best if I directly ask now:  Is that the case?  Any thoughts, etc. you can share there?  I know you're being 'paid' for your contributions to Marlin via Patreon and directly (as well you should, I contributed too, as you know), but is that the direction you'd like to go permanently?  Any insight you can share would be helpful to me - and maybe others - I think.

To be open, I'm really having fun working on this project, but am not sure how I feel about that kind of direction at the moment (not that I should have much say, as someone relatively new, but I certainly do decide how I allocate my time).

> I've been taking on some of that role in handling the Twitter for @MarlinFirmware, writing a lot of the website, managing releases, and writing semi-regular updates on my Patreon page, but haven't formally adopted an executive role in the project itself.

Just so I understand how things are organized/structured, generally:  Do you feel it is / should be your right to adopt an 'executive role' (in any form)?  I want it to be clear that there's nothing implied or suggested in that question - I just want to understand how you see things.

> For the time being, please let them ferment a bit longer. 

That's totally fine.  I had thought you'd already looked over things when you provided commentary previously, so didn't think it would be a problem to merge, given that they're all 'safe' (tested, conditionally compiled) changes, and for the reasons I gave above.

> I very much want to move forward with 1.2, the new file hierarchy, and the HAL. And I might have even wanted to hold off on this new non-essential feature until after then. But I'm waiting for things to slow down in the issue queue and make sure that 1.1.x is solid before leaping forward with 1.2.

I was hoping that you were going to move forward on the 1.2 changes soon - and as you know I offered to help with the changes directly, to stop working on new things, etc. - but when there were no responses to the last thread on this topic (that I'm aware of), no group discussion like we'd considered, no indicated direction, no proposed timeframe communicated, etc. I assumed there was no need to worry with any of it, and that when you were ready to freeze things and move forward you would announce it.  No sarcasm intended, but I (we) can't tell what you're thinking, and in the absence of information, I'm going to act in what I think are the best interests of the project, which is what I did here.

> Perhaps with our combined energy we can patch the bugs, fix the issues, clean up the code, and make sure everything works nicely while holding back new features and contributions and putting off everyone so that we can have a nice slow point and make the transition to 1.2.

That sounds great - I very much think those things need to be done, and I love what I understand of what's coming - but there's been no timeframe proposed except in general/vague terms, no feature freeze date announced, etc.  If you want to lead that effort, then propose a timeframe and I'm sure others can chime in and we can come to a consensus and act accordingly.

Phew!  I didn't intend for this to be that long, but I guess there was a lot to say.  Thanks for taking the time to read this.

bgort commented:
> I've moved the encoder firmware to a standalone repository here. Hopefully that makes the code a tad easier to work with - I can understand not wanting to pull a repo full of CAD files to tweak a few lines of code. If there's anything else I can do to make it easier to work with just let me know.

Great!  Will do.

> Happy to see any PRs for bugs you find, from what you describe it sounds like the WDT isn't resetting the modules and I'm not sure what would be causing that. There have been a few changes to the code since I last thoroughly tested things, but nothing (I thought) that touched on that area of things.

I'll fork the new repo, take some more time to see what I can find, etc. and will do a PR if I have any suggestions.  I glanced earlier but nothing stood out.

> Just want to say I really appreciate the work you've put in cleaning this up and getting it integrated into Marlin nicely! 

Absolutely.  I've been enjoying working on this - for whatever reason it's exciting to me - and look forward to working on it even more.  

> Seeing the progress you were making was my motivation to start fleshing out the v2 modules, so many thanks for that. Hopefully, soon there will be some new hardware to test your improvements with!

That's one of the things I love about working on projects like this -- motivation compounds/builds due to the efforts of others when you have the right mix of people.  In that same vein, I would likely be motivated to work to implement any new features you might include in the new boards, too.

> My supplier for one-off / small quantities of the magnetic strips during testing was through AMS here. AFAIK they're just short pre-cut lengths of a strip from Bogen Electronic. They have a datasheet that covers most of the strip's specifications on that page, though I'm not sure if they specifically state the accuracy class (they seem to list a few options).

Thanks.  I'll check in with them directly.  300mm is fine for all of my axes but one.

> Strip with 1mm poles / 2mm pole-pair length is all that I've tried, as it seems to be what's recommended in the datasheet. I've gotten pole-length and pole-pair length mixed up a few times, so all I'd say is double check which numbers are being talked about - I didn't realise @jaysonkelly and co were actually using a strip of a different density, but if 1024 ticks/mm is working for them then they must be.

Yeah, the datasheet is pretty clear about non-linearities/error when using other lengths, so I'd probably prefer use what they recommend.  @jaysonkelly says it works fine and I already have his version, so I'll try it out, however.  Will report here with what I learn.

> Anyway, let me know if there's anything else I can do to help here. Otherwise, I am watching keenly!

Absolutely!  Thanks for all of your hard work, too.  I really like that you're including the regulator on the v2, and the extra connectors for daisy-chaining.

bgort commented:
> One thing I would certainly have mentioned, had this been held longer, is that you cannot use N as a G-code parameter. You'll have to change M864 to use another letter. Please also avoid M and G.

Ah, shoot.  I'll move the `N` to `W`.  The sacredness of the gcode (as Roxy would say) is still a bit lost on me, it seems.

> @bgort We will also need to make I2CPositionEncoder into a static singleton by making all its data and methods static.

We can't do that right now, as there are multiple instances of I2CPositionEncoder - one for each encoder.  I was considering consolidating I2CPositionEncoder with I2CPositionEncodersMgr (which I already had plans to make static), and if I do that I can probably make them all static.  Still lots to do.

thinkyhead commented:
> this is what I expected - a followup PR that made the corrections you felt were necessary

And it is what you got…. #7007 — still inert and pending.

> I2CPositionEncodersMgr

Only this is now `static`. `I2CPositionEncoder` is restored to non-static.
